### PR TITLE
#161967136 favouriting articles 

### PR DESCRIPTION
**Pull request by denomash**
**What does this PR do?**
-  favorites an article

**Description of Task to be completed?**
- authenticated users can favorite an article
- authenticated users can unfavorite an article
- authenticated users can view a list of favorited articles

**How should this be manually tested?**

- Clone the repo and cd into the project folder
- Create a virtual environment by running `python3 -m venv <name_of_environment>`
- Activate the virtual environment by running `source <name_of_environment>/bin/activate`
- Run `$ pip install -r requirements.txt`
- Then add the following environment variables to your computer:

 - `$ export DB_NAME='choose_a_dbname'`
 - `$ export DB_USER='db_username'`
 - `$ export DB_PASSWORD='db_role_password'`
 - `$ export DB_HOST='localhost or preffered host'`
 - `$ export DB_PORT='server_port_number'`
- `$ export SECRET_KEY='some_secret_key`
- `$ export SOCIAL_AUTH_FACEBOOK_KEY='<facebook_app_id>'`
- `$ export SOCIAL_AUTH_FACEBOOK_SECRET='<facebook_app_secret>'`
- `$ export SOCIAL_AUTH_TWITTER_KEY='<twitter_api_key>'`
- `$ export SOCIAL_AUTH_TWITTER_SECRET='<twitter_api_secret>'`
- `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_KEY='<google_client_id>'`
- `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET='<google_client_secret>'`

- Change the current branch by running `git checkout  ft-favouriting-articles-161967136`
- Run `python3 manage.py makemigrations` then, `python3 manage.py migrate`
- Run `python3 manage.py runserver`

- To favorite an article, use the following endpoint:
**GET** `article/<str:slug>/favorite`

- To unfavorite an article , use the following endpoint:
**DEL** `article/<str:slug>/favorite`

- To get all unfavorited articles, use the following endpoint:
**GET** `articles/all/favorites`

**What are the relevant pivotal tracker stories?**
- [#61967136](https://www.pivotaltracker.com/story/show/161967136/comments/197727307)

**Checklist (edited)**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

**Screenshots (if appropriate)**
**Favorite an article**
screenshot 2018-12-18 at 16 03 18

**A flag is raised if article has been favorited**
screenshot 2018-12-18 at 16 03 18

**Unfavorite an article**
screenshot 2018-12-18 at 16 05 40

**A flag is raised if article has been unfavorited**
screenshot 2018-12-18 at 16 05 47

**Get all favorited articles**
screenshot 2018-12-18 at 16 09 14
screenshot 2018-12-18 at 16 10 33

**Get article by slug**
screenshot 2018-12-18 at 16 11 31

### CODE REVIEW

**Review by davishooly — APPROVED** [review]

File: authors/apps/favorite/views.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        favorite = FavouriteArticle.objects.filter(user=self.request.user.id, article=res.id).exists()
+        if favorite:
+            return Response({"message": "You have already favorited this article"}, status=status.HTTP_400_BAD_REQUEST)
+
+        data = {"user": request.user.id, "article": res.id}
+        serializer = self.serializer_class(data=data)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+
+        message = {"message": "You have favorited this article"}
```
Comment (bettblake08) [inline authors/apps/favorite/views.py]: For all your messages, define them in the response.py file and call them in your views and tests.

**Review by bettblake08 — CHANGES_REQUESTED** [review]

File: authors/apps/favorite/views.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    renderer_classes = (FavoriteJSONRenderer,)
+
+    def post(self, request, slug):
+        """
+        Favorite a specific article
+        :param slug
+        authenticated users can favorite an article
+        """
+
+        res = check_exist(slug)
```
Comment (bettblake08) [inline authors/apps/favorite/views.py]: Have a more descriptive variable name for the res variable. Its not python standard

File: authors/apps/favorite/views.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+        if isinstance(res, Response):
+            return res
+
+        # check if user has favourite an article
+        # If user has favourited the article raise a flag
+        # otherwise favourite the article
+        favorite = FavouriteArticle.objects.filter(user=self.request.user.id, article=res.id).exists()
+        if favorite:
+            return Response({"message": "You have already favorited this article"}, status=status.HTTP_400_BAD_REQUEST)
```
Comment (bettblake08) [inline authors/apps/favorite/views.py]: You need to have a hanging indentation for this line of code. It's too long.

**Review by bettblake08 — CHANGES_REQUESTED** [review]

**Comment by denomash** [comment]
@bettblake08 I implemented on the areas you had highlighted

**Review by Bryan-Cee — APPROVED** [review]
👍

**Review by bettblake08 — APPROVED** [review]
👍 It's good

**Review by Bryan-Cee — CHANGES_REQUESTED** [review]
Check on coverage

**Review by Bryan-Cee — APPROVED** [review]

**Review by mutoromike — APPROVED** [review]
LGTM

**Pull request merged**